app.js

```diff
@@ -1,36 +1,144 @@
-/**
- * Copyright 2017, Google, Inc.
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *    http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
 
 'use strict';
 
-// [START gae_node_request_example]
 const express = require('express');
+const cookieParser = require('cookie-parser');
+const uniqueString = require('unique-string');
+
+const logger = require('./logger');
+
+// TODO: none of this setup is pretty
+const {CloudStorage, MemoryStorage} = require('./storage');
+const storage = process.env.NODE_ENV === 'production'
+   ? new CloudStorage
```

Comment:
So in prod, test reports are stored in GCS? Is there an expiration/lifecycle? Otherwise, is this a DoS concern?

Reply:
The bucket (mdn-bcd-buffer) has a "15 days or older" retention policy. It could be much shorter, but this is enough to ensure it doesn't grow forever. It would still be possible to spam the bucket of course, but it's not possible to read it back except as pull requests on a GitHub repository, so it would be for troublemaking and not for free file hosting.

One could require authentication, but that would make it more work to test on odd browsers. For example, I ran it on a phone in a store once, and I wouldn't do that if it took any work.